Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,315 @@
+            "writeData returned \(dataLength) bytes, but actually \(originalDataEndIndex - originalDataStartIndex) bytes were written. They must be the same."
+        )
+
+        // We write the length after the data to begin with. We will move it later
+
+        /// The actual number of bytes used to write the length written. The user may write more or fewer bytes than what we reserved
+        let actualIntegerLength = strategy.writeInteger(
+            dataLength,
+            reservedCapacity: reservedCapacity,
+            to: &self
+        )
+
+        switch actualIntegerLength {
+        case reservedCapacity:
+            // Good, exact match, swap the values and then "delete" the trailing bytes by moving the index back
+            self._moveBytes(from: originalDataEndIndex, to: lengthPrefixIndex, size: actualIntegerLength)
+            self.moveWriterIndex(to: originalDataEndIndex)
+        case ..<reservedCapacity:
+            // We wrote fewer bytes. We now have to move the length bytes from the end, and
+            // _then_ shrink the rest of the buffer onto it.
+            self._moveBytes(from: originalDataEndIndex, to: lengthPrefixIndex, size: actualIntegerLength)
+            let newDataStartIndex = lengthPrefixIndex + actualIntegerLength
+            self._moveBytes(
+                from: originalDataStartIndex,
+                to: newDataStartIndex,
+                size: dataLength
+            )
+            self.moveWriterIndex(to: newDataStartIndex + dataLength)
+        case reservedCapacity...:
+            // We wrote more bytes. We now have to create enough space. Once we do, we have the same
+            // implementation as the matching case.
+            let extraSpaceNeeded = actualIntegerLength - reservedCapacity
+            self._createSpace(before: lengthPrefixIndex, requiredSpace: extraSpaceNeeded)
+
+            // Clean up the indices.
+            let newDataEndIndex = originalDataEndIndex + extraSpaceNeeded
+            // We wrote the length after the data, so we have to move those bytes to the space at the front
+            self._moveBytes(from: newDataEndIndex, to: lengthPrefixIndex, size: actualIntegerLength)
+            self.moveWriterIndex(to: newDataEndIndex)
+        default:
+            fatalError("Unreachable")
+        }
+
+        let totalBytesWritten = self.writerIndex - lengthPrefixIndex
+        return totalBytesWritten
+    }
+
+    /// Reads a slice which is prefixed with a length. The length will be read using `strategy`, and then that many bytes will be read to create a slice.
+    /// - Returns: The slice, if there are enough bytes to read it fully. In this case, the readerIndex will move to after the slice.
+    /// If there are not enough bytes to read the full slice, the readerIndex will stay unchanged.
+    @inlinable
+    public mutating func readLengthPrefixedSlice<Strategy: NIOBinaryIntegerEncodingStrategy>(
+        _ strategy: Strategy
```

Comment:
Not sure about dropping the label on the `strategy` here, I think we should keep the strategy label.